File: app/src/main/java/com/akrapovic/soundkit/community/domain/QuietWindowEvaluator.kt
package com.akrapovic.soundkit.community.domain

import java.util.Calendar
import java.util.TimeZone

object QuietWindowEvaluator {
    fun isOvernight(quiet: QuietStartSettings): Boolean =
        quiet.windowEndMinute < quiet.windowStartMinute

    fun isActive(
        quiet: QuietStartSettings,
        now: Calendar = Calendar.getInstance(TimeZone.getDefault()),
    ): Boolean {
        if (!quiet.enabled) return false
        val dayOfWeek = (now.get(Calendar.DAY_OF_WEEK) + 5) % 7
        if (dayOfWeek !in quiet.daysOfWeek) return false
        val minuteOfDay = now.get(Calendar.HOUR_OF_DAY) * 60 + now.get(Calendar.MINUTE)
        return if (quiet.windowEndMinute >= quiet.windowStartMinute) {
            minuteOfDay in quiet.windowStartMinute..quiet.windowEndMinute
        } else {
            minuteOfDay >= quiet.windowStartMinute || minuteOfDay <= quiet.windowEndMinute
        }
    }
}

File: ios/SoundKitCommunity/Domain/QuietWindowEvaluator.swift
import Foundation

enum QuietWindowEvaluator {
    static func isOvernight(_ quiet: QuietStartSettings) -> Bool {
        quiet.windowEndMinute < quiet.windowStartMinute
    }

    static func isActive(_ quiet: QuietStartSettings, now: Date = Date()) -> Bool {
        guard quiet.enabled else { return false }
        var calendar = Calendar.current
        calendar.timeZone = .current
        let weekday = (calendar.component(.weekday, from: now) + 5) % 7
        guard quiet.daysOfWeek.contains(weekday) else { return false }
        let hour = calendar.component(.hour, from: now)
        let minute = calendar.component(.minute, from: now)
        let minuteOfDay = hour * 60 + minute
        if quiet.windowEndMinute >= quiet.windowStartMinute {
            return minuteOfDay >= quiet.windowStartMinute && minuteOfDay <= quiet.windowEndMinute
        }
        return minuteOfDay >= quiet.windowStartMinute || minuteOfDay <= quiet.windowEndMinute
    }
}
